docs/GENERATORS.md

```diff
@@ -0,0 +1,102 @@
+
+## Adding A Complex Generator
+
+A more *complex* generator would be the ComplexNumbers generator found [here](https://github.com/exercism/csharp/blob/master/generators/Exercises/ComplexNumbers.cs).
+
+The `Exercise` abstract class currently exposes five methods that are used for overriding the default behavior when generating an exercise.
+
+### void UpdateCanonicalData(CanonicalData canonicalData)
+Update the canonical data for a given test. 
+
+The most common use for this override is to iterate over each of the canonical data cases.
+
+As an example, if you wanted to change the default behavior so that when the `Input` value of a test is a negative number, an exception should be thrown, the code would look like this.
+
+```csharp
+protected override void UpdateCanonicalData(CanonicalData canonicalData)
+{
+  foreach (var canonicalDataCase in canonicalData.Cases)
+  {
+    var caseInputLessThanZero = (long)canonicalDataCase.Input["number"] < 0;
+    canonicalDataCase.ExceptionThrown = caseInputLessThanZero ? typeof(ArgumentException) : null;
+  }
+}
+```
+
+### HashSet\<string\> AddAdditionalNamespaces()
+Allows more namespaces to be added to the test suite. 
+
+The tests use `Xunit` so all tests will automatically include the `Xunit` namespace. However, more advanced tests may require additional namespaces.
+
+```csharp
+protected override HashSet<string> AddAdditionalNamespaces()
+{
+  return new HashSet<string>()
+  {
+    typeof(Dictionary<char, int>).Namespace
+  };
+}
+```
+
+This snippet would add the namespace that the `Dictionary<char,int>` collection lives in (`System.Collections.Generic`).
+
+### string RenderTestMethodBody[Arrange/Act/Assert]
+Override the default behavior when rendering a test methods arrange, act, and/or assert sections.
+
+More advanced tests may need to leverage a `template`. A template allows you to add additional code to a test and assert more complex statements.
+
+An example of this is the [RunLengthEncoding](https://github.com/exercism/csharp/blob/master/generators/Exercises/RunLengthEncoding.cs) test.
+
+Here the **Assert** is being overridden. The assert needs to call additional functions, but only if the property is `consistency`. Otherwise, render the assert as usual.
+
+## Updating Existing Files
+It is possible that an existing exercise does not match the canonical data. It is OK to update the exercise stub and/or the exercise example to follow the canonical data! An example might be that an exercise is named SumOfMultiples, but the SumOfMultiples.cs and Example.cs files both use `Multiples` as the name of the class.
+
+Also, if you find an issue with one of the existing generators or test suites simply open up the generator that you would like to update, make your changes, and then run the generators.
+
+## Running The Generators
```

Comment:
Is there a way currently to run one one exercise? I don't think so, but if so, it's worth calling out.

Reply:
@robkeim There is, by using the `-e <exercise>` command-line parameter.

Reply:
Nice! I think that's worth calling out in the md then.